lib/Dialect/TensorExt/Transforms/RotateAndReduce.cpp

```diff
@@ -58,167 +59,17 @@ struct RotateAndReduce : impl::RotateAndReduceBase<RotateAndReduce> {
-                arith::ConstantOp indexConstant =
-                    tensorOp.getIndices()
-                        .front()
-                        .template getDefiningOp<arith::ConstantOp>();
-                if (!indexConstant) {
-                  LLVM_DEBUG(
-                      llvm::dbgs()
-                      << "Not replacing op due to non constant index access;"
-                      << " (do you need to run --canonicalize or --sccp?)\n");
-                  return failure();
-                }
-                int64_t accessIndex =
-                    mlir::cast<IntegerAttr>(indexConstant.getValue()).getInt();
-
-                // If the access index was already seen, then fail because some
-                // tensor element contributes more than once to the reduction.
-                if (accessIndices.count(accessIndex)) {
-                  LLVM_DEBUG(
-                      llvm::dbgs()
-                      << "Not replacing op because input tensor was accessed "
-                         "multiple times in at same index\n");
-                  return failure();
-                }
-                LLVM_DEBUG(llvm::dbgs()
-                           << "Adding valid index " << accessIndex << "\n");
-                accessIndices.insert(accessIndex);
-                return success();
-              })
-              .Default([&](Operation *op) {
-                LLVM_DEBUG(llvm::dbgs() << "Not continuing because type switch "
-                                           "encountered unsupported op "
-                                        << op->getName() << "\n");
-                return failure();
-              });
-
-      if (failed(result)) {
-        return;
-      }
-    }
-
-    // The test for a match is now: does the number of accessed indices exactly
-    // match the size of the tensor? I.e., does each tensor element show up
-    // exactly once in the reduction?
-    if (inputTensors.empty()) {
-      LLVM_DEBUG(llvm::dbgs()
-                 << "Not replacing op because it accesses no tensors\n");
-      return;
-    }
-    auto tensorShape =
-        mlir::cast<RankedTensorType>(inputTensors.begin()->getType())
-            .getShape();
-    if (tensorShape.size() != 1 || tensorShape[0] != accessIndices.size()) {
-      LLVM_DEBUG(llvm::dbgs()
-                 << "Not replacing op because tensor shape ("
-                 << inputTensors.begin()->getType()
-                 << ") is not fully reduced. Only " << accessIndices.size()
-                 << " of " << tensorShape[0] << " indices were accessed\n");
-      return;
+    if (extraction) {
+      // We can extract at any index; every index contains the same reduced
+      // value.
+      auto extractOp = b.create<tensor::ExtractOp>(
+          finalOp->getResult(0),
+          b.create<arith::ConstantIndexOp>(0).getResult());
+      op->replaceAllUsesWith(extractOp);
+    } else {
+      op->replaceAllUsesWith(finalOp);
```

Comment:
ah, interesting - could you add a testcase for this branch and/or a comment on line 99 that the extraction boolean adds a final extraction operation for the sum? 

```
// Sum all entries of a tensor
// CHECK-LABEL: @simple_sum_mixed_rotation_tensor
// CHECK-SAME: (%[[arg0:.*]]: tensor<8xi32>
// CHECK-NEXT: %[[c0:.*]] = arith.constant 0
// CHECK-NEXT: %[[c1:.*]] = arith.constant 1
// CHECK-NEXT: %[[c2:.*]] = arith.constant 2
// CHECK-NEXT: %[[c4:.*]] = arith.constant 4
// CHECK-NEXT: %[[v0:.*]] = tensor_ext.rotate %[[arg0]], %[[c4]]
// CHECK-NEXT: %[[v1:.*]] = arith.addi %[[arg0]], %[[v0]]
// CHECK-NEXT: %[[v2:.*]] = tensor_ext.rotate %[[v1]], %[[c2]]
// CHECK-NEXT: %[[v3:.*]] = arith.addi %[[v1]], %[[v2]]
// CHECK-NEXT: %[[v4:.*]] = tensor_ext.rotate %[[v3]], %[[c1]]
// CHECK-NEXT: %[[v5:.*]] = arith.addi %[[v3]], %[[v4]]
// CHECK-NEXT: return %[[v5]]
func.func @simple_sum_mixed_rotation_tensor(%arg0: tensor<8xi32>) -> tensor<8xi32> {
  %c0 = arith.constant 0 : index
  %c1 = arith.constant 1 : index
  %c2 = arith.constant 2 : index
  %c3 = arith.constant 3 : index
  %c4 = arith.constant 4 : index
  %c5 = arith.constant 5 : index
  %c6 = arith.constant 6 : index
  %c7 = arith.constant 7 : index
  %0 = tensor_ext.rotate %arg0, %c1 : tensor<8xi32>, index
  %1 = tensor_ext.rotate %arg0, %c2 : tensor<8xi32>, index
  %2 = arith.addi %0, %1 : tensor<8xi32>
  %3 = tensor_ext.rotate %arg0, %c3 : tensor<8xi32>, index
  %4 = tensor_ext.rotate %arg0, %c4 : tensor<8xi32>, index
  %5 = tensor_ext.rotate %arg0, %c5 : tensor<8xi32>, index
  %6 = tensor_ext.rotate %arg0, %c6 : tensor<8xi32>, index
  %7 = tensor_ext.rotate %arg0, %c7 : tensor<8xi32>, index
  %8 = arith.addi %2, %3 : tensor<8xi32>
  %9 = arith.addi %8, %4 : tensor<8xi32>
  %10 = arith.addi %9, %5 : tensor<8xi32>
  %11 = arith.addi %10, %6 : tensor<8xi32>
  %12 = arith.addi %11, %7 : tensor<8xi32>
  %13 = arith.addi %12, %arg0 : tensor<8xi32>
  return %13 : tensor<8xi32>
}
```

Reply:
Both added

Reply:
thank you!